src/js/game/hud/parts/buildings_toolbar.js

```diff
@@ -109,20 +109,39 @@ export class HUDBuildingsToolbar extends BaseHUDPart {
     }
 
     cycleBuildings() {
+        const actionMapper = this.root.keyMapper;
         let newIndex = this.lastSelectedIndex;
-        for (let i = 0; i < toolbarBuildings.length; ++i, ++newIndex) {
-            newIndex %= toolbarBuildings.length;
-            const metaBuilding = gMetaBuildingRegistry.findByClass(toolbarBuildings[newIndex]);
-            const handle = this.buildingHandles[metaBuilding.id];
-            if (!handle.selected && handle.unlocked) {
-                break;
+
+        if (actionMapper.getBinding(KEYMAPPINGS.placementModifiers.cycleInverse).isCurrentlyPressed()) {
```

Comment:
I'm a bit confused here. Why don't you simply make a variable `direction` (which is -1 or 1 depending if its shift or not). Then you add that to `this.lastSelectedIndex` until you find a valid building or reach the beginning. I feel like that would be much more intuitive (Thinking about readability)

Reply:
You're right, that would be more readable - apparently i already did it with the variant cycling. 

I simplified the `cycleBuildings()` code with commit 6266406.